reconcile/aus/base.py

```diff
@@ -95,72 +105,187 @@ def get_ocm_env_upgrade_specs(
         ...
 
 
-# consider first lower versions and lower soakdays (when versions are equal)
-def sort_key(d: dict) -> tuple:
-    return (
-        parse_semver(d["current_version"]),
-        d["conditions"].get("soakDays") or 0,
-    )
+class GateAgreement(BaseModel):
+    id: str
+
+    def create(self, ocm: OCM, cluster_name: str) -> None:
+        action_log(
+            "create",
+            ocm.name,
+            cluster_name,
+            f"Creating version agreement for gate {self.id}",
+        )
+        agreement = ocm.create_version_agreement(self.id, cluster_name)
+        if agreement.get("version_gate") is None:
+            logging.error(
+                "Unexpected response while creating version "
+                f"agreement with id {self.id} for cluster {cluster_name}"
+            )
+
+
+class AbstractUpgradePolicy(ABC, BaseModel):
+    """Abstract class for upgrade policies
+    Used to create and delete upgrade policies in OCM."""
+
+    cluster: str
+    gates_to_agree: Optional[list[GateAgreement]]
+    id: Optional[str]
+    next_run: Optional[str]
+    schedule: Optional[str]
+    schedule_type: str
+    version: str
+
+    @abstractmethod
+    def create(self, ocm: OCM) -> None:
+        pass
+
+    @abstractmethod
+    def delete(self, ocm: OCM) -> None:
+        pass
+
+    def create_gate_agreements(self, ocm: OCM) -> None:
```

Comment:
This should only apply to cluster upgrades, not to addon upgrades. So it should probably be in the `ClusterUpgradePolicy` class, right ? At least the implementation that actually does something. If you want to keep an abstract method at this level, the `AddonUpgradePolicy` implementation should do nothing

Reply:
Good point, I moved this to ClusterUpgradePolicy.